src/cmdstan/stansummary_helper.hpp

```diff
@@ -190,31 +190,434 @@ int matrix_index(std::vector<int> &index, const std::vector<int> &dims) {
+                                     Eigen::VectorXd &warmup_times,
+                                     Eigen::VectorXd &sampling_times,
+                                     Eigen::VectorXi &thin, std::ostream *out) {
+  // instantiate stan::mcmc::chains object by parsing first file
+  std::ifstream ifstream;
+  ifstream.open(filenames[0].c_str());
+  stan::io::stan_csv stan_csv = stan::io::stan_csv_reader::parse(ifstream, out);
+  warmup_times(0) = stan_csv.timing.warmup;
+  sampling_times(0) = stan_csv.timing.sampling;
+  stan::mcmc::chains<> chains(stan_csv);
+  ifstream.close();
+  thin(0) = stan_csv.metadata.thin;
+  metadata = stan_csv.metadata;
+
+  // parse rest of input files, add to chains
+  for (std::vector<std::string>::size_type chain = 1; chain < filenames.size();
+       chain++) {
+    ifstream.open(filenames[chain].c_str());
+    stan_csv = stan::io::stan_csv_reader::parse(ifstream, out);
+    chains.add(stan_csv);
+    ifstream.close();
+    thin(chain) = stan_csv.metadata.thin;
+    warmup_times(chain) = stan_csv.timing.warmup;
+    sampling_times(chain) = stan_csv.timing.sampling;
+  }
+  return chains;
+}
+
+std::vector<std::string> get_sampler_params_header(
+    const std::vector<std::string> &percentiles) {
+  // Mean, StdDev, ... percentiles ...
+  std::vector<std::string> sampler_params_header(percentiles.size() + 2);
+  sampler_params_header.at(0) = "Mean";
+  sampler_params_header.at(1) = "StdDev";
+  size_t offset = 2;
+  for (size_t i = 0; i < percentiles.size(); ++i) {
+    sampler_params_header[i + offset] = percentiles[i] + '%';
+  }
+  return sampler_params_header;
+}
+
+std::vector<std::string> get_model_params_header(
+    const std::vector<std::string> &percentiles) {
+  // Mean, MCSE,  StdDev, ... percentiles ..., N_eff, N_eff/s, R_hat
+  std::vector<std::string> model_params_header(percentiles.size() + 6);
+  model_params_header.at(0) = "Mean";
+  model_params_header.at(1) = "MCSE";
+  model_params_header.at(2) = "StdDev";
+  size_t offset = 3;
+  for (size_t i = 0; i < percentiles.size(); ++i) {
+    model_params_header[i + offset] = percentiles[i] + '%';
+  }
+  offset += percentiles.size();
+  model_params_header.at(offset) = "N_Eff";
+  model_params_header.at(offset + 1) = "N_Eff/s";
+  model_params_header.at(offset + 2) = "R_hat";
+  return model_params_header;
+}
+
+void sampler_params_stats(const stan::mcmc::chains<> &chains,
+                          const Eigen::VectorXd &probs,
+                          int sampler_params_start_col,
+                          Eigen::MatrixXd &sampler_params) {
+  sampler_params.setZero();
+  for (int i = 0; i < sampler_params.rows(); ++i) {
+    int i_offset = i + sampler_params_start_col;
+    sampler_params(i, 0) = chains.mean(i_offset);
+    sampler_params(i, 1) = chains.sd(i_offset);
+    Eigen::VectorXd quantiles = chains.quantiles(i_offset, probs);
+    for (int j = 0; j < probs.size(); j++)
+      sampler_params(i, 2 + j) = quantiles(j);
+  }
+}
```

Comment:
Am I right that this is checking the cumulative mean, sd, and quantiles of the sampler parameter statistics?

It would be a good sanity check that `sampler_params.size()` is equal to `2 + probs.size()`

Reply:
skipping sanity check as this isn't coming from user, it's all internal program logic.